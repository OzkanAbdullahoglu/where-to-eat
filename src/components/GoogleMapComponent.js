import React from 'react';

/* global google*/
import {
  withScriptjs,
  withGoogleMap,
  GoogleMap,
  Marker,
} from 'react-google-maps';
import { compose, withProps, lifecycle } from 'recompose';
import { connect } from 'react-redux';
import { ActivityIndicator } from 'react-native-web';
import { getFetchedData } from '../reducers';

let filteredFetchedData;
/**
 *  @description creating GoogleMap API structure and implement it to the DOM
 *  react-google-maps is used for implementation
 */
const GoogleMapComponent = compose(
  withProps({
    googleMapURL:
      'https://maps.googleapis.com/maps/api/js?libraries=geometry,places&key=AIzaSyDt0R1NakICIfNoUjNBjQEU3BlIr_QhtHY&v=3.exp',
    loadingElement: <div style={{ height: '100%' }} />,
    containerElement: <div className="container-element" />,
    mapElement: <div className="map-element" />,
  }),
  lifecycle({
    componentDidCatch(error) {
      alert(
        `${error} Error Occured while trying to render google maps API Please check your credentials`
      );
    },
  }),
  withScriptjs,
  withGoogleMap
)((props) => {
  const markerLinkRedirection = (url) =>
    window.open(url, '_blank');

  if (!props.isFetchedData) {
    return (
      <ActivityIndicator color="#2E3B4B" />
    );
  }

  filteredFetchedData = props.isFetchedData.filter((data) => data.coordinates !== null);

  return (
    <GoogleMap
      defaultZoom={11}
      defaultCenter={{ lat: 41.0082376, lng: 28.9783589 }}
      center={props.center}
      zoom={props.zoom}
    >
      {/*
        mapping through updatedLocations array
        we implement markers with their lat and lng information
        we assigin all markers an unique key as an id of venue
      */}
      {/*
        we animate clicked marker with Animation.BOUNCE
        we locate exact clicked marker with its unique key
      */}
      {/*
        with onClick event here we trigger
        handleToggle for implementation of InfoWindow
        implementPhoto to trigger photo fetching process from FourSquare API
        markerOnTriggered to find out which marker was clicked
        zoomTrigger to update center of map to the marker's exact location and to update zoom
      */}

      {filteredFetchedData.map((getUpdatedLoc) => (
        <Marker
          position={{
            lat: getUpdatedLoc.coordinates.latitude,
            lng: getUpdatedLoc.coordinates.longitude,
          }}
          key={getUpdatedLoc.id}
          animation={
            props.markerUniqueKey === getUpdatedLoc.id &&
            props.markerClicked === true &&
            google.maps.Animation.BOUNCE
          }
          onClick={() => markerLinkRedirection(getUpdatedLoc.url)}
        />
      ))}
    </GoogleMap>
  );
});

const mapStateToProps = (store) => ({
  isFetchedData: getFetchedData(store),
});

const withRedux = connect(mapStateToProps);

export default compose(withRedux)(GoogleMapComponent);

