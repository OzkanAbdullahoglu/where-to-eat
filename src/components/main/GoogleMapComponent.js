/* eslint-disable linebreak-style */
import React from 'react';

/* global google*/
import {
  withScriptjs,
  withGoogleMap,
  GoogleMap,
  Marker,
} from 'react-google-maps';
import { compose, withProps, lifecycle } from 'recompose';
import { connect } from 'react-redux';
import { getFetchedData } from '../../reducers';

let filteredFetchedData;
/**
 *  @description creating GoogleMap API structure and implement it to the DOM
 *  react-google-maps is used for implementation
 */
const GoogleMapComponent = compose(
  withProps({
    googleMapURL:
      'https://maps.googleapis.com/maps/api/js?libraries=geometry,places&key=AIzaSyDt0R1NakICIfNoUjNBjQEU3BlIr_QhtHY&v=3.exp',
    loadingElement: <div style={{ height: '100%' }} />,
    containerElement: <div className="container-element" />,
    mapElement: <div className="map-element" />,
  }),
  lifecycle({
    componentDidCatch(error) {
      alert(
        `${error} Error Occured while trying to render google maps API Please check your credentials`
      );
    },
    shouldComponentUpdate(nextProps) {
      const { center, zoom, isFetchedData } = this.props;
      return (center !== nextProps.center ||
        zoom !== nextProps.zoom ||
        isFetchedData !== nextProps.isFetchedData);
    },
  }),
  withScriptjs,
  withGoogleMap
)((props) => {
  const markerLinkRedirection = (url) =>
    window.open(url, '_blank');

  if (props.isFetchedData === '') {
    return (
      <GoogleMap
        defaultZoom={11}
        center={props.center}
        zoom={props.zoom}
      />
    );
  }

  filteredFetchedData = props.isFetchedData.filter((data) => data.coordinates !== null);
  return (
    <GoogleMap
      defaultZoom={11}
      center={props.center}
      zoom={props.zoom}
    >
      {/*
        mapping through updatedLocations array
        we implement markers with their lat and lng information
        we assigin all markers an unique key as an id of venue
      */}
      {/*
        we animate clicked marker with Animation.BOUNCE
        we locate exact clicked marker with its unique key
      */}
      {/*
        with onClick event here we trigger
        handleToggle for implementation of InfoWindow
        implementPhoto to trigger photo fetching process from FourSquare API
        markerOnTriggered to find out which marker was clicked
        zoomTrigger to update center of map to the marker's exact location and to update zoom
      */}

      {filteredFetchedData.map((getUpdatedLoc) => (
        <Marker
          position={{
            lat: getUpdatedLoc.coordinates.latitude,
            lng: getUpdatedLoc.coordinates.longitude,
          }}
          key={getUpdatedLoc.id}
          onClick={() => markerLinkRedirection(getUpdatedLoc.url)}
        />
      ))}
    </GoogleMap>
  );
});

const mapStateToProps = (store) => ({
  isFetchedData: getFetchedData(store),
});

const withRedux = connect(mapStateToProps);

export default compose(withRedux)(GoogleMapComponent);

