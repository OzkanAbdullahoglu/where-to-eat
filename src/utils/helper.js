
export const openHoursManage = (obj) => {
  let willOpen = '';
  let willClosed = '';
  let openUntil = '';
  const date = new Date();
  const today = date.getDay();
  const time = parseInt(
    `${JSON.stringify(date.getHours())}${JSON.stringify(date.getMinutes())}`, 10
  );

  if (obj === undefined || obj.open[today] === undefined) {
    return '';
  }
  const closeTimeInt = parseInt(obj.open[today].end, 10);
  const openTime = `${obj.open[today].start.slice(0, 2)}:${obj.open[today].start.slice(2)}`;
  const closeTime = `${obj.open[today].end.slice(0, 2)}:${obj.open[
    today
  ].end.slice(2)}`;
  if (closeTimeInt - time < 100) {
    willClosed = `Close soon ${closeTime}`;
  } else if (time > closeTimeInt || time < closeTimeInt) {
    willOpen = `Opens at ${openTime}`;
  } else {
    openUntil = `Open until ${closeTime}`;
  }
  return (
    willClosed || willOpen || openUntil
  );
};

export const getCategoriesTitle = (arr) => {
  const titleArr = arr.map((obj) => obj.title);
  const concatenateTitles = titleArr.reduce(
    (accTitles, singleTitle) => `${accTitles}, ${singleTitle}`
  );
  return concatenateTitles;
};
