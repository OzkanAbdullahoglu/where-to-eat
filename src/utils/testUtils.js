/* eslint-disable linebreak-style */
import { createStore, applyMiddleware } from 'redux';
import { middlewares } from '../store';
import rootReducer from '../reducers';


export const storeFactory = (initalState) => {
  const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
  return createStoreWithMiddleware(rootReducer, initalState);
}
;
