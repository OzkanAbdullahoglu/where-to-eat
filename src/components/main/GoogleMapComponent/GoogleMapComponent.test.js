/* eslint-disable linebreak-style */
import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from '@testing-library/react';
import toJson from 'enzyme-to-json';
import { GoogleMapComponent } from './GoogleMapComponent';
import { storeFactory } from '../../../utils/testUtils';

describe('GoogleMapComponent', () => {
  const store = storeFactory();
  jest.mock('./GoogleMapComponent', () => function DummyMap(props) {
    return (
      <div data-testid="map">
        {props.center.lat}:{props.center.long}
      </div>
    );
  });
  let googleMapComponent;

  beforeEach(() => {
    act(() => {
      googleMapComponent = mount(
        <GoogleMapComponent store={store} />
      );
    });
  });

  afterEach(() => {
    googleMapComponent.unmount();
  });

  it('renders GoogleMapComponent component correctly', () => {
    expect(toJson(googleMapComponent)).toMatchSnapshot();
  });
});

function orderMeal(isHungry) {
  if (isHungry) {
    const toDoWhenIsHungry = "Order today's specials";
  } else {
    const toDoWhenIsNotHungry = 'Naah! I would like to order a drink.';
    console.log(toDoWhenIsHungry);
  }
}
