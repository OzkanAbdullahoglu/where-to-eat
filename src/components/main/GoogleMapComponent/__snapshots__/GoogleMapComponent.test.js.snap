// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleMapComponent renders GoogleMapComponent component correctly 1`] = `
<withProps(lifecycle(withScriptjs(withGoogleMap(Component))))
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <lifecycle(withScriptjs(withGoogleMap(Component)))
    containerElement={
      <div
        className="container-element"
      />
    }
    googleMapURL="https://maps.googleapis.com/maps/api/js?libraries=geometry,places&key=AIzaSyDt0R1NakICIfNoUjNBjQEU3BlIr_QhtHY&v=3.exp"
    loadingElement={
      <div
        style={
          Object {
            "height": "100%",
          }
        }
      />
    }
    mapElement={
      <div
        className="map-element"
      />
    }
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <withScriptjs(withGoogleMap(Component))
      containerElement={
        <div
          className="container-element"
        />
      }
      googleMapURL="https://maps.googleapis.com/maps/api/js?libraries=geometry,places&key=AIzaSyDt0R1NakICIfNoUjNBjQEU3BlIr_QhtHY&v=3.exp"
      loadingElement={
        <div
          style={
            Object {
              "height": "100%",
            }
          }
        />
      }
      mapElement={
        <div
          className="map-element"
        />
      }
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <div
        style={
          Object {
            "height": "100%",
          }
        }
      />
    </withScriptjs(withGoogleMap(Component))>
  </lifecycle(withScriptjs(withGoogleMap(Component)))>
</withProps(lifecycle(withScriptjs(withGoogleMap(Component))))>
`;
